system/Helpers/array_helper.php

```diff
@@ -122,3 +122,81 @@ function array_deep_search($key, array $array)
 		return null;
 	}
 }
+
+if (!function_exists('sort_by_multiple_keys'))
+{
+	/**
+	 * Sorts an associative multidimensional array by the values of the
+	 * passed columns names. The sorting params represent an associative
+	 * array of column names and sorting flags.
+	 * Both arrays of objects and arrays of array can be sorted.
+	 *
+	 * Example:
+	 * 	sort_by_multiple_keys($players,
+	 * 			[
+	 * 				'team.hierarchy' => SORT_ASC,
+	 * 				'position'       => SORT_ASC,
+	 * 				'name'           => SORT_STRING,
+	 * 			]
+	 * 		);
+	 * The '.' dot operator in the column name indicates a deeper array or
+	 * object level. In principle, any number of sublevels could be used,
+	 * as long as the level and column exist in every array element.
+	 *
+	 * @param array $array the reference of the array to be sorted
+	 * @param array $sortColumns an associative array of columns to sort
+	 *              after and their sorting flags
+	 *
+	 * @return boolean
+	 */
+	function array_sort_by_multiple_keys(array &$array, array $sortColumns): bool
+	{
+		// Check if there really are columns to sort after
+		if (empty($sortColumns) || empty($array))
+		{
+			return false;
+		}
+
+		// Group sorting indexes and data
+		$tempArray = [];
+		foreach ($sortColumns as $key => $sortFlag)
+		{
+			// Get sorting values
+			$carry = $array;
+
+			// The '.' operator separates nested elements
+			foreach (explode('.', $key) as $keySegment)
+			{
+				// Distinguish array of objects and arrays of arrays
+				if (is_object(reset($carry)) ?? false)
+				{
+					// Extract the object attribute
+					foreach ($carry as $index => $object)
+					{
+						$carry[$index] = $object->$keySegment;
+					}
+				}
+				else
+				{
+					// Extract
+					$carry = array_column($carry, $keySegment);
+				}
+			}
+
+			// Store this iterations sorting paramters
+			$tempArray[$key][0] = $carry;
+			$tempArray[$key][1] = $sortFlag;
+		}
+
+		// Prepare sorting array
+		$sortArray = [];
+		foreach ($sortColumns as $index => $sortFlag)
+		{
+			$sortArray[] = &$tempArray[$index][0];
+			$sortArray[] = &$tempArray[$index][1];
+		}
+		$sortArray[] = &$array;
+
+		return call_user_func_array('array_multisort', $sortArray);
```

Comment:
Why did you use `call_user_func_array` instead of calling `array_multisort` directly?

Reply:
To enable sorting with multiple priority levels  `array_multisort` needs to receive each of the value arrays followed by the respective sort flag(s) in a single call. As it only accepts a list, this was the solution I came up with.
If we just call `array_multisort` in each of the `foreach` loops with a single 'value array' and flag, then in the end the `&$array` will only be sorted by the last entry in `$sortColumns`.

Reply:
> If we just call `array_multisort` in each of the `foreach` loops with a single 'value array' and flag, then in the end the `&$array` will only be sorted by the last entry in `$sortColumns`.

But I see it called *after* the `foreach` loop, not during. Am I missing something?

Reply:
Its called after `foreach`, because we want it sorted by all of the data we collected in the `foreach `at once.

See `Example #3` here:
https://www.php.net/manual/de/function.array-multisort.php

We need to feed a list of onedimensional arrays and sorting flags to `array_multisort`. In our case, we don't know the length of this list, as it depends on the `$sortColumns` that the user passes. Therefore, we collect all arguments in a single array and feed it to `array_multisort `via `call_user_func_array`, which translates the `$tempArray` into a list of function arguments.
I hope this makes it clearer :)

Reply:
Ah I get it now. You are unpacking the arguments to `array_multisort`. Maybe we can have a small comment there explaining the need for the `call_user_func_array` so that future refactors won't mess with that.

Thanks for the thorough explanation, though. 😁

Reply:
Alright :D Will push that with the tests to save on phpunit time.

Reply:
Its in.